### DISCUSSION TITLE
Is first context special?

### DISCUSSION THREAD

**Discussion by sorenchr2011** [inducer/pycuda#430] [Q&A]
Hi,
 I have an issue when creating a second a pycuda context (GL enabled in my case).
Here's a minimal example - functions are at the bottom of the post.
I create OpenGL and pycuda contexts and transfer texture data using pycuda. This works as expected.
I then take down the pycuda context (pop), set up a new one and repeat the above - this fails with a segfault. Interestingly, if the variable of first pycuda context is still in scope (the one deactivated with .pop()), then the below works.
So it looks like the pycuda calls depend on the first context - why is this, what am I misunderstanding? Why can't I pop one context and start a new one?
Thanks!
Soren

using pycuda==2023.1

```
device_index = 0
texture_fill_value = 101
pycuda.driver.init()
opengl_context, texture_id = setup_open_gl_context()
cuda_context = setup_cuda_context()

test_opengl_to_cuda_interop(texture_id)
test_opengl_to_cuda_interop(texture_id) # still works..
cuda_context.pop()

# now if I do  :   'del cuda_context' here, then the below will segfault inside test_opengl_to_cuda_interop at  pycuda.gl.RegisteredImage

cuda_context1 = setup_cuda_context()
test_opengl_to_cuda_interop(texture_id)   #this fails if the original context was destroyed
cuda_context1.pop()


```

```


def  setup_open_gl_context():
    ctx = moderngl.create_context(standalone=True,
                              backend='egl',
                              device_index=device_index)
    tex = ctx.texture(
        size=(256, 256), data=np.full((256, 256), dtype=np.float16, fill_value=texture_fill_value),
        components=1,
        dtype='f2')
    return ctx, tex.glo

def setup_cuda_context():
    dev = pycuda.driver.Device(device_index)
    pycuda_context = pycuda.gl.make_context(dev, flags=0)
    return pycuda_context

def test_opengl_to_cuda_interop(texture_id):
    tensor = torch.zeros([256, 256, 1], dtype=torch.float16, device=torch.device('cuda:0'))
    cuda_buffer = pycuda.gl.RegisteredImage(texture_id, GL_TEXTURE_2D, pycuda.gl.graphics_map_flags.NONE)
    cuda_buffer_map = cuda_buffer.map()
    h, w, c = tensor.shape
    copy_op = pycuda.driver.Memcpy2D()
    copy_op.set_src_array(cuda_buffer_map.array(0, 0))
    copy_op.set_dst_device(tensor.data_ptr())
    copy_op.width_in_bytes = w * c * tensor.element_size()
    copy_op.src_pitch = copy_op.width_in_bytes
    copy_op.dst_pitch = copy_op.width_in_bytes
    copy_op.height = tensor.shape[0]
    copy_op(aligned=False)
    np.testing.assert_allclose(tensor.cpu().numpy(), texture_fill_value, atol=0.01)
    cuda_buffer_map.unmap()
    cuda_buffer.unregister()

```

**Comment by inducer** [comment]
It looks like the neither context is not set up to do GL interop. See [here](https://documen.tician.de/pycuda/gl.html#pycuda.gl.make_context). Attempting interop from that context is not set up for interop is, at best, undefined behavior, IMO.

**Comment by sorenchr2011** [comment]
Thanks - but sorry I don't understand that - isn't that what pycuda_context = pycuda.gl.make_context(dev, flags=0) should be doing? (in setup_cuda_context())
That is what you are linking too also - what am I missing?

**Reply by inducer** [reply]
Oh, I missed that. The other question is whether more than one context can interoperate with a given GL context at a time. I do not know the answer to this, but I suspect it might be "no". I've certainly not seen it done. Maybe Nvidia's developer documentation can provide a definitive answer.
